sql/tests: TestRandomSyntaxGeneration failed [`ALTER TENANT (SELECT 'foo') START REPLICATION ...` immediately crashes the database process]

sql/tests.TestRandomSyntaxGeneration [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/17949844?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_RandomSyntaxTestsBazel/17949844?buildTab=artifacts#/) on master @ [f2ce52ff7cc3b4680e0fec2ef844baad49042d6a](https://github.com/cockroachdb/cockroach/commits/f2ce52ff7cc3b4680e0fec2ef844baad49042d6a):


```
*   | 	pkg/sql/sem/tree/type_check.go:345
*   | github.com/cockroachdb/cockroach/pkg/sql/sem/tree.TypeCheckAndRequire
*   | 	pkg/sql/sem/tree/type_check.go:355
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).analyzeExpr
*   | 	pkg/sql/analyze_expr.go:34
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).planTenantSpec
*   | 	pkg/sql/tenant_spec.go:63
*   | github.com/cockroachdb/cockroach/pkg/sql.(*planner).LookupTenantInfo
*   | 	pkg/sql/tenant_spec.go:148
*   | github.com/cockroachdb/cockroach/pkg/ccl/crosscluster/physical.alterReplicationJobHook.func1
*   | 	pkg/ccl/crosscluster/physical/alter_replication_job.go:220
*   | github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1
*   | 	pkg/sql/planhook.go:194
*   | github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2
*   | 	pkg/util/stop/stopper.go:498
*   | runtime.goexit
*   | 	src/runtime/asm_amd64.s:1695
* Wraps: (4) ReturnType called on TypedExpr with empty typeAnnotation. Was the underlying Expr type-checked before asserting a type of TypedExpr?
* Error types: (1) *withstack.withStack (2) *assert.withAssertionFailure (3) *withstack.withStack (4) *errutil.leafError
*
panic: ReturnType called on TypedExpr with empty typeAnnotation. Was the underlying Expr type-checked before asserting a type of TypedExpr? [recovered]
	panic: ReturnType called on TypedExpr with empty typeAnnotation. Was the underlying Expr type-checked before asserting a type of TypedExpr?

goroutine 1330694 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0x454aa5?, {0x89cace8, 0xc00cac0e70})
	pkg/util/stop/stopper.go:226 +0x65
panic({0x69852e0?, 0xc00c0ab010?})
	GOROOT/src/runtime/panic.go:770 +0x132
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.typeAnnotation.assertTyped(...)
	pkg/sql/sem/tree/expr.go:148
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.(*Subquery).TypeCheck(0x42e198?, {0x0?, 0x0?}, 0x0?, 0x8f00000000000000?)
	pkg/sql/sem/tree/type_check.go:1764 +0xd3
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.TypeCheck({0x89cace8?, 0xc00cac0e70?}, {0x89cbb90?, 0xc00cac06f0?}, 0x706dda6?, 0xc001518660?)
	pkg/sql/sem/tree/type_check.go:345 +0x8a
github.com/cockroachdb/cockroach/pkg/sql/sem/tree.TypeCheckAndRequire({0x89cace8?, 0xc00cac0e70?}, {0x89cbb90?, 0xc00cac06f0?}, 0xc004b92c00?, 0xca8d740, {0x71334f7, 0x21})
	pkg/sql/sem/tree/type_check.go:355 +0x54
github.com/cockroachdb/cockroach/pkg/sql.(*planner).analyzeExpr(0xc00d434138, {0x89cace8, 0xc00cac0e70}, {0x89cbb90?, 0xc00cac06f0?}, {{0x0, 0x0, 0x0}, {0x0, 0x0}}, ...)
	pkg/sql/analyze_expr.go:34 +0xaf
github.com/cockroachdb/cockroach/pkg/sql.(*planner).planTenantSpec(0xc00cac0e70?, {0x89cace8?, 0xc00cac0e70?}, 0x1e?, {0x71334f7?, 0x4000000000000000?})
	pkg/sql/tenant_spec.go:63 +0x190
github.com/cockroachdb/cockroach/pkg/sql.(*planner).LookupTenantInfo(0xc00d434138, {0x89cace8, 0xc00cac0e70}, 0x89a81c0?, {0x71334f7?, 0xc004b92e10?})
	pkg/sql/tenant_spec.go:148 +0x27
github.com/cockroachdb/cockroach/pkg/ccl/crosscluster/physical.alterReplicationJobHook.func1({0x89cace8, 0xc00cac0e70}, {0x30000c00610e640?, 0xc004b92e90?, 0xd61c74?}, 0xc031b05500)
	pkg/ccl/crosscluster/physical/alter_replication_job.go:220 +0x159
github.com/cockroachdb/cockroach/pkg/sql.(*hookFnNode).startExec.func1({0x89cace8, 0xc00cac0e70})
	pkg/sql/planhook.go:194 +0x4a
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx.func2({0x70d0130?, 0x0?})
	pkg/util/stop/stopper.go:498 +0x1f0
created by github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunAsyncTaskEx in goroutine 1327598
	pkg/util/stop/stopper.go:488 +0x47c
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
<details><summary>Same failure on other branches</summary>
<p>

- #136274 sql/tests: TestRandomSyntaxGeneration failed [C-test-failure O-robot T-sql-foundations branch-release-24.3 release-blocker]
- #135784 sql/tests: TestRandomSyntaxGeneration failed [Alter PCR Stream] [A-disaster-recovery C-test-failure O-robot T-disaster-recovery branch-release-24.2 release-blocker]
- #132558 sql/tests: TestRandomSyntaxGeneration failed [CREATE TENANT statement is not valid after parsing then formatting] [C-test-failure O-robot P-2 T-db-server branch-release-24.1.6-rc]
</p>
</details>
/cc @cockroachdb/sql-foundations
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestRandomSyntaxGeneration.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/bazci/githubpost/issues)

</sub>


Jira issue: CRDB-44975

--- a/pkg/crosscluster/physical/alter_replication_job.go
+++ b/pkg/crosscluster/physical/alter_replication_job.go
@@ -200,6 +200,12 @@ func alterReplicationJobHook(
 		}
 	}
 
+	// Ensure the TenantSpec is type checked, even if we don't use the result.
+	_, _, _, err = exprEval.TenantSpec(ctx, alterTenantStmt.TenantSpec)
+	if err != nil {
+		return nil, nil, nil, false, err
+	}
+
 	retentionTTLSeconds := defaultRetentionTTLSeconds
 	if ret, ok := options.GetRetention(); ok {
 		retentionTTLSeconds = ret

--- a/pkg/crosscluster/physical/alter_replication_job_test.go
+++ b/pkg/crosscluster/physical/alter_replication_job_test.go
@@ -597,3 +597,22 @@ func TestAlterTenantStartReplicationAfterRestore(t *testing.T) {
 	srcTime := srv.Clock().Now()
 	replicationtestutils.WaitUntilReplicatedTime(t, srcTime, db, catpb.JobID(ingestionJobID))
 }
+
+func TestAlterReplicationJobErrors(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+	defer log.Scope(t).Close(t)
+
+	ctx := context.Background()
+
+	srv, sqlDB, _ := serverutils.StartServer(t, base.TestServerArgs{
+		DefaultTestTenant: base.TestControlsTenantsExplicitly})
+	defer srv.Stopper().Stop(ctx)
+
+	db := sqlutils.MakeSQLRunner(sqlDB)
+
+	t.Run("alter tenant subqueries", func(t *testing.T) {
+		// Regression test for #136339
+		db.ExpectErr(t, "subqueries are not allowed", "ALTER TENANT (select 't2') START REPLICATION OF t1 ON 'foo'")
+	})
+
+}

